### DISCUSSION TITLE
integrating-vite-with-asp-net-core-a-winning-combination/

### DISCUSSION THREAD

**Discussion by martijnboland** [martijnboland/blog#58] [Blog comments]
# integrating-vite-with-asp-net-core-a-winning-combination/

https://blogs.taiga.nl/martijn/posts/integrating-vite-with-asp-net-core-a-winning-combination/

**Comment by disqus-comments-migrator** [comment]
<em>[sadegh](https://disqus.com/by/sadegh051/) commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at juni 20 2021, 06:43</em>

---

<p>Nice But How to Vite + Asp + Vue ?<br>and how to handle vite-legacy</p>

**Reply by martijnboland** [reply]
<em>[Martijn Boland](https://disqus.com/by/martijnboland/) commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at juni 21 2021, 09:36, in reply to [sadegh](https://disqus.com/by/sadegh051/)'s [comment](https://github.com/martijnboland/blog/discussions/58#discussioncomment-8953080)</em>

---

<p>Vite + ASP.NET Core and Vue will work the same as with React. Don't know for vite-legacy because I haven't needed it yet.</p>

**Reply by disqus-comments-migrator** [reply]
<em>[sadegh](https://disqus.com/by/sadegh051/) commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at juli 04 2021, 10:35, in reply to [Martijn Boland](https://disqus.com/by/martijnboland/)'s [comment](https://github.com/martijnboland/blog/discussions/58#discussioncomment-8953082)</em>

---

<p>what about SpaServices ?</p><p><b>Microsoft.AspNetCore.SpaServices.ReactDevelopmentServer</b> will work with vue ?</p>

**Reply by martijnboland** [reply]
<em>[Martijn Boland](https://disqus.com/by/martijnboland/) commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at juli 05 2021, 10:45, in reply to [sadegh](https://disqus.com/by/sadegh051/)'s [comment](https://github.com/martijnboland/blog/discussions/58#discussioncomment-8953084)</em>

---

<p>Yes, I guess it will work because in fact, SpaServices.ReactDevelopmentServer has actually nothing to do with react. It just starts a nodejs development server from an npm script.</p><p>One thing: make sure that the development server outputs "Starting the development server" because the SpaServices middleware checks this to ensure the development server is running. In the example in this post, this is done in the "start" script in package.json file:</p><p><code><br>"scripts": {<br>    "start": "echo Starting the development server &amp;&amp; vite"<br>}<br><br></code></p>

**Comment by disqus-comments-migrator** [comment]
<em>[Niels Vanhorenbeeck](https://disqus.com/by/disqus_lBAuAqQ3DA/) commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at januari 27 2022, 03:59</em>

---

<p>Hello, I followed your tutorial to migrate from WebPack to Vite and it almost works, thank you for that. Production bundling is very fast. However, during development, it seems Vite recompiles everything on every page load (multi page Dot Net app). So it takes more than 10 seconds to render the page after every click. I also don't seem to find the reason for this, any idea?</p>

**Reply by martijnboland** [reply]
<em>[Martijn Boland](https://disqus.com/by/martijnboland/) commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at januari 27 2022, 07:06, in reply to [Niels Vanhorenbeeck](https://disqus.com/by/disqus_lBAuAqQ3DA/)'s [comment](https://github.com/martijnboland/blog/discussions/58#discussioncomment-8953087)</em>

---

<p>When running the example app locally, I also sometimes notice recompiling (unstyled content for a few seconds). This was not happening before the latest upgrade to .NET6 and the latest vite. Can't find the exact reason, but at I'm going to find out if it's a Vite issue or .NET6. Stay tuned :-)</p>

**Reply by disqus-comments-migrator** [reply]
<em>[Niels Vanhorenbeeck](https://disqus.com/by/disqus_lBAuAqQ3DA/) commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at januari 28 2022, 09:11, in reply to [Martijn Boland](https://disqus.com/by/martijnboland/)'s [comment](https://github.com/martijnboland/blog/discussions/58#discussioncomment-8953089)</em>

---

<p>Im working in .NET5. Can be mistaken, but without the part in Startup.cs, it still recompiles every page load, but much faster. Still not nearly enough though.</p>

**Reply by martijnboland** [reply]
<em>[Martijn Boland](https://disqus.com/by/martijnboland/) commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at januari 30 2022, 11:49, in reply to [Niels Vanhorenbeeck](https://disqus.com/by/disqus_lBAuAqQ3DA/)'s [comment](https://github.com/martijnboland/blog/discussions/58#discussioncomment-8953090)</em>

---

<p>I may have found the issue: ASP.NET Core is proxying requests to http://localhost:3000. On Windows, localhost is resolved to IPV6 ::1 first and then after 2 seconds or so to 127.0.0.1. When explicitly setting the host option in package.json to listen to all addresses, things got fast again ("start": "echo Starting the development server &amp;&amp; vite --host ::").<br>The stupid thing however is that with old Vite versions, there were no problems. Perhaps newer Vite versions don't listen on ::1 by default anymore?</p>

**Comment by disqus-comments-migrator** [comment]
<em>[linus ang](https://disqus.com/by/linusang/) commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at februari 03 2022, 05:34</em>

---

<p>does this work with frontend routing? for example if we have a url that looks like this "<a href="http://domain.com/about/team" rel="nofollow noopener" title="domain.com/about/team">domain.com/about/team</a>", "/about" will be a razor page and "/team" will be a vue-router page.. is this possible?</p>

**Reply by martijnboland** [reply]
<em>[Martijn Boland](https://disqus.com/by/martijnboland/) commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at februari 04 2022, 10:35, in reply to [linus ang](https://disqus.com/by/linusang/)'s [comment](https://github.com/martijnboland/blog/discussions/58#discussioncomment-8953092)</em>

---

<p>Yes, that's possible. The key here is to configure ASP.NET so that requests to /about/team render the about razor page. This can be done easily via the @page directive: @page "/about/{**anything}". You can also set this up for all razor pages via a convention: <a href="https://docs.microsoft.com/en-us/aspnet/core/razor-pages/razor-pages-conventions?view=aspnetcore-6.0" rel="nofollow noopener" title="https://docs.microsoft.com/en-us/aspnet/core/razor-pages/razor-pages-conventions?view=aspnetcore-6.0">https://docs.microsoft.com/en-us/aspnet/core/razor-pages/razor-pages-conventions?view=aspnetcore-6.0</a>.<br> Perhaps you have to add /about to all vue-router routes, or maybe it's possible to add a base path?</p>

**Comment by disqus-comments-migrator** [comment]
<em>[Simon Boddy](https://disqus.com/by/simonboddy/) commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at mei 31 2022, 10:31</em>

---

<p>This is golden information. I was right with you up til the "final step". You ask me to put something in my Razor page. I don't have a razor page, don't want one, and don't understand why you have one. Neither do I have a wwroot folder. I'm working with a fresh <code>dotnet new react</code>, from which I deleted and recreated the ClientApp using <code>npm init vite@latest</code>. So I'm successfully clicking play in VS, getting redirected from some high port to localhost:3000, and seeing my Vue landing page. Will I be able to access my apis? I shall report back.</p>

**Reply by martijnboland** [reply]
<em>[Martijn Boland](https://disqus.com/by/martijnboland/) commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at mei 31 2022, 10:50, in reply to [Simon Boddy](https://disqus.com/by/simonboddy/)'s [comment](https://github.com/martijnboland/blog/discussions/58#discussioncomment-8953094)</em>

---

<p>From what you're describing it looks like you're building a 100% Single Page App. I think you should just be able to access the api's just like in the react template. The things for the Razor page are especially for scenario's where you have a mixed server-side/SPA application. I think you can just skip that part in your case.</p>

**Reply by disqus-comments-migrator** [reply]
<em>[Simon Boddy](https://disqus.com/by/simonboddy/) commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at mei 31 2022, 03:24, in reply to [Martijn Boland](https://disqus.com/by/martijnboland/)'s [comment](https://github.com/martijnboland/blog/discussions/58#discussioncomment-8953096)</em>

---

<p>This proved to be more of a rabbit hole than I was expecting! <code>dotnet new react</code> gives you 2 .js files that spit out environment variables for your https cert and key. These are then read by webpack to give you https in development. To get this to work with Vite, you need to copy aspnetcore-https.js and aspnetcore-react.js into the root of your Vite app, then substantially modify vite.config.ts. <br>It's a bit of a fiddle, but I learnt a ton about node and environment variables, and as you say, the result promises to be well worth it.</p><p>My working vite.config.ts is...</p><p><code><br>import { defineConfig, loadEnv } from 'vite'<br>import vue from '@vitejs/plugin-vue'<br>import * as fs from 'node:fs'<br><br>// https://vitejs.dev/config/<br>export default ({ mode }) =&gt; {<br>    process.env = { ...process.env, ...loadEnv(mode, process.cwd(),'') };<br>    console.log(process.env);<br>    return defineConfig({<br>        plugins: [vue()],<br>        base: '/dist/',<br>        server: {<br>            hmr: {<br>                protocol: 'wss'<br>            },<br>            port: 5002,<br>            https: {<br>                key: fs.readFileSync(process.env.SSL_KEY_FILE),<br>                cert: fs.readFileSync(process.env.SSL_CRT_FILE)<br>            },<br>            proxy: {<br>              '/api': {<br>                target: 'https://localhost:5001/',<br>                changeOrigin: true,<br>                secure: false<br>              }<br>            }<br>        }<br>    })<br>}</code></p>

**Reply by martijnboland** [reply]
<em>[Martijn Boland](https://disqus.com/by/martijnboland/) commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at mei 31 2022, 04:20, in reply to [Simon Boddy](https://disqus.com/by/simonboddy/)'s [comment](https://github.com/martijnboland/blog/discussions/58#discussioncomment-8953097)</em>

---

<p>Alright! Thanks for the heads up. This is probably only required when running the ASP.NET dev server with SSL, right?</p>

**Reply by disqus-comments-migrator** [reply]
<em>[Simon Boddy](https://disqus.com/by/simonboddy/) commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at mei 31 2022, 08:37, in reply to [Martijn Boland](https://disqus.com/by/martijnboland/)'s [comment](https://github.com/martijnboland/blog/discussions/58#discussioncomment-8953099)</em>

---

<p>Yup. A lazy person could just change the devserver url to http, but I wanted to stay as close to the dotnet react template as possible. I found <a href="https://github.com/johanndev/asp_vite" rel="nofollow noopener" title="https://github.com/johanndev/asp_vite">this github repo</a> that demonstrates a working project.</p>

**Comment by disqus-comments-migrator** [comment]
<em>[Luke Pinion](https://disqus.com/by/luke_pinion/) commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at september 15 2022, 05:34</em>

---

<p>Thanks for this great tutorial! I'm still struggling with getting it up and running on our end as the structure of our project is different than yours, but I think I'll get there!</p><p>At any rate, I wanted to advise you that the code snippets for your Razor pages seem to be broken. It appears the HTML entities aren't getting converted correctly.</p>

**Reply by martijnboland** [reply]
<em>[Martijn Boland](https://disqus.com/by/martijnboland/) commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at september 16 2022, 05:23, in reply to [Luke Pinion](https://disqus.com/by/luke_pinion/)'s [comment](https://github.com/martijnboland/blog/discussions/58#discussioncomment-8953102)</em>

---

<p>Thanks for the hint. I'll see if I can fix this (used to be fine in the past). Meanwhile I wish you good luck with getting it up and running!</p>

**Comment by disqus-comments-migrator** [comment]
<em>[Rodrigo Bonvin de Andrade](https://disqus.com/by/rodrigobonvindeandrade/) commented [on Disqus](https://disqus.com/home/forum/martijnboland/) at september 26 2022, 09:22</em>

---

<p>Tanks man!!</p>
